### ISSUE TITLE
Problem with Amazon IGV backend setup

### ISSUE THREAD

**Issue by pelacables** [igvteam/igv#764]
Hi all,

I'll trying to configure IGV so it can read files from a S3 bucket and I'm following the guide in https://umccr.org/blog/igv-amazon-backend-setup/.
I managed to configure cognito and I could login, but, when I try to load a file from the bucket I get some errors that I don't fully understand (I believ the problem is located in the fetch token process, but I'm not sure, the message say very few things)..

I see 3 errors:
after the login (the browser says "OK" after the login process):

```bash
ERROR [2020-02-27T12:26:27,390] [OAuthProvider.java:240]  java.lang.NullPointerException
```

The second is when I try to browse the buscket:

```bash
ERROR [2020-02-27T12:26:55,282] [DefaultExceptionHandler.java:49]  Unhandled exception
java.lang.NullPointerException: null
```

and the last one when I specify the s3 url:

```bash
ERROR [2020-02-27T12:26:55,282] [DefaultExceptionHandler.java:49]  Unhandled exception
java.lang.NullPointerException: null
```

Here's the full igv.log:

```bash
INFO [2020-02-27T12:25:47,378] [Main.java:159]  Startup  IGV Version 2.8.0 01/06/2020 08:43 PM
INFO [2020-02-27T12:25:47,388] [Main.java:160]  Java 11.0.5 (build 11.0.5+10) 2019-10-15
INFO [2020-02-27T12:25:47,389] [Main.java:163]  Java Vendor: AdoptOpenJDK https://adoptopenjdk.net/
INFO [2020-02-27T12:25:47,390] [Main.java:165]  JVM: OpenJDK 64-Bit Server VM AdoptOpenJDK   
INFO [2020-02-27T12:25:47,390] [Main.java:168]  Default User Directory: \\XYZ\bria\Documents
INFO [2020-02-27T12:25:47,391] [Main.java:169]  OS: Windows 10 10.0 amd64
INFO [2020-02-27T12:25:47,807] [AmazonUtils.java:68]  AWS configuration found. AWS support enabled under 'Amazon' menu
INFO [2020-02-27T12:25:47,971] [GenomeManager.java:192]  Loading genome: C:\Users\bria\igv\genomes\hg19.genome
INFO [2020-02-27T12:25:54,628] [GenomeComboBox.java:90]  Enter genome combo box
INFO [2020-02-27T12:25:54,646] [GenomeManager.java:284]  Genome loaded.  id= hg19
INFO [2020-02-27T12:25:54,659] [CommandListener.java:120]  Listening on port 60151
INFO [2020-02-27T12:26:26,414] [CommandListener.java:186]  GET /oauthCallback?code=136fbf97-3a40-4d67-93b3-565546c0507a&state=abadf119-1dec-48f1-90a5-35bd185320c1 HTTP/1.1
ERROR [2020-02-27T12:26:27,390] [OAuthProvider.java:240]  java.lang.NullPointerException
INFO [2020-02-27T12:26:27,454] [CommandListener.java:186]  GET /favicon.ico HTTP/1.1
ERROR [2020-02-27T12:26:55,282] [DefaultExceptionHandler.java:49]  Unhandled exception
java.lang.NullPointerException: null
	at org.broad.igv.util.AmazonUtils.GetCognitoAWSCredentials(AmazonUtils.java:100) ~[igv.jar:?]
	at org.broad.igv.util.AmazonUtils.ListBucketsForUser(AmazonUtils.java:191) ~[igv.jar:?]
	at org.broad.igv.ui.IGVMenuBar.lambda$createAWSMenu$6(IGVMenuBar.java:972) ~[igv.jar:?]
	at javax.swing.AbstractButton.fireActionPerformed(Unknown Source) ~[?:?]
	at javax.swing.AbstractButton$Handler.actionPerformed(Unknown Source) ~[?:?]
	at javax.swing.DefaultButtonModel.fireActionPerformed(Unknown Source) ~[?:?]
	at javax.swing.DefaultButtonModel.setPressed(Unknown Source) ~[?:?]
	at javax.swing.AbstractButton.doClick(Unknown Source) ~[?:?]
	at com.jidesoft.plaf.vsnet.VsnetMenuItemUI.doClick(VsnetMenuItemUI.java:1395) ~[jide-common-3.7.3.jar:?]
	at com.jidesoft.plaf.vsnet.VsnetMenuItemUI$MouseInputHandler.mouseReleased(VsnetMenuItemUI.java:1190) ~[jide-common-3.7.3.jar:?]
	at java.awt.Component.processMouseEvent(Unknown Source) ~[?:?]
	at javax.swing.JComponent.processMouseEvent(Unknown Source) ~[?:?]
	at java.awt.Component.processEvent(Unknown Source) ~[?:?]
	at java.awt.Container.processEvent(Unknown Source) ~[?:?]
	at java.awt.Component.dispatchEventImpl(Unknown Source) ~[?:?]
	at java.awt.Container.dispatchEventImpl(Unknown Source) ~[?:?]
	at java.awt.Component.dispatchEvent(Unknown Source) ~[?:?]
	at java.awt.LightweightDispatcher.retargetMouseEvent(Unknown Source) ~[?:?]
	at java.awt.LightweightDispatcher.processMouseEvent(Unknown Source) ~[?:?]
	at java.awt.LightweightDispatcher.dispatchEvent(Unknown Source) ~[?:?]
	at java.awt.Container.dispatchEventImpl(Unknown Source) ~[?:?]
	at java.awt.Window.dispatchEventImpl(Unknown Source) ~[?:?]
	at java.awt.Component.dispatchEvent(Unknown Source) ~[?:?]
	at java.awt.EventQueue.dispatchEventImpl(Unknown Source) ~[?:?]
	at java.awt.EventQueue$4.run(Unknown Source) ~[?:?]
	at java.awt.EventQueue$4.run(Unknown Source) ~[?:?]
	at java.security.AccessController.doPrivileged(Native Method) ~[?:?]
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source) ~[?:?]
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source) ~[?:?]
	at java.awt.EventQueue$5.run(Unknown Source) ~[?:?]
	at java.awt.EventQueue$5.run(Unknown Source) ~[?:?]
	at java.security.AccessController.doPrivileged(Native Method) ~[?:?]
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(Unknown Source) ~[?:?]
	at java.awt.EventQueue.dispatchEvent(Unknown Source) ~[?:?]
	at java.awt.EventDispatchThread.pumpOneEventForFilters(Unknown Source) [?:?]
	at java.awt.EventDispatchThread.pumpEventsForFilter(Unknown Source) [?:?]
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(Unknown Source) [?:?]
	at java.awt.EventDispatchThread.pumpEvents(Unknown Source) [?:?]
	at java.awt.EventDispatchThread.pumpEvents(Unknown Source) [?:?]
	at java.awt.EventDispatchThread.run(Unknown Source) [?:?]
INFO [2020-02-27T12:27:55,742] [IGV.java:1338]  Loading 1 resources.
INFO [2020-02-27T12:27:55,746] [TrackLoader.java:126]  Loading resource, path s3://bucket/de7e8cc8-7485-40c1-9ec4-961cbd206c8e_star_genome.mark_dups.bam
ERROR [2020-02-27T12:27:55,746] [TrackLoader.java:253]  null
java.lang.NullPointerException: null
	at org.broad.igv.util.AmazonUtils.GetCognitoAWSCredentials(AmazonUtils.java:100) ~[igv.jar:?]
	at org.broad.igv.util.AmazonUtils.createPresignedURL(AmazonUtils.java:309) ~[igv.jar:?]
	at org.broad.igv.util.AmazonUtils.translateAmazonCloudURL(AmazonUtils.java:340) ~[igv.jar:?]
	at org.broad.igv.util.HttpUtils.mapURL(HttpUtils.java:133) ~[igv.jar:?]
	at org.broad.igv.util.HttpUtils.createURL(HttpUtils.java:122) ~[igv.jar:?]
	at org.broad.igv.util.ResourceLocator.getTypeString(ResourceLocator.java:192) ~[igv.jar:?]
	at org.broad.igv.util.ResourceLocator.getTypeString(ResourceLocator.java:173) ~[igv.jar:?]
	at org.broad.igv.track.TrackLoader.load(TrackLoader.java:128) [igv.jar:?]
	at org.broad.igv.ui.IGV.load(IGV.java:1421) [igv.jar:?]
	at org.broad.igv.ui.IGV.loadResources(IGV.java:1353) [igv.jar:?]
	at org.broad.igv.ui.IGV$4.run(IGV.java:474) [igv.jar:?]
	at org.broad.igv.util.LongRunningTask.call(LongRunningTask.java:72) [igv.jar:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
ERROR [2020-02-27T12:27:55,747] [IGV.java:1356]  Error loading track
java.lang.NullPointerException: null
	at org.broad.igv.exceptions.DataLoadException.<init>(DataLoadException.java:38) ~[igv.jar:?]
	at org.broad.igv.track.TrackLoader.load(TrackLoader.java:255) ~[igv.jar:?]
	at org.broad.igv.ui.IGV.load(IGV.java:1421) ~[igv.jar:?]
	at org.broad.igv.ui.IGV.loadResources(IGV.java:1353) [igv.jar:?]
	at org.broad.igv.ui.IGV$4.run(IGV.java:474) [igv.jar:?]
	at org.broad.igv.util.LongRunningTask.call(LongRunningTask.java:72) [igv.jar:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
INFO [2020-02-27T12:27:55,750] [MessageUtils.java:76]  Error loading s3://bucket/de7e8cc8-7485-40c1-9ec4-961cbd206c8e_star_genome.mark_dups.bam: null
```

Anyone with more experience could help me?

TIA,
Arnau

**Comment by jrobinso** [comment]
@brainstorm Any ideas?

Here's the code in context, the NPE is on the last line.  My guess is response.get("id_token") is returning null.     

```
OAuthProvider provider = OAuthUtils.getInstance().getProvider("Amazon");
JsonObject igv_oauth_conf = GetCognitoConfig();
JsonObject response = provider.getResponse();
JsonObject payload = JWTParser.getPayload(response.get("id_token").getAsString());
```

**Comment by jrobinso** [comment]
@pelacables would it be possible to grant us access and let us see the configuration file?  Or could you setup a test case that would be sharable?   You can send info privately to igv-team (at) broadinstitute.org.

**Comment by brainstorm** [comment]
@pelacables Which IGV version are you using right now? Can we take a look at your oauth-config.json? It should look somewhat like this:

```json
{
   "apiKey": "",
   "project_id": "igv",

   "auth_provider": "Amazon",
   "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
   "aws_region": "ap-southeast-2",
   "scope": "email%20openid%20profile",
   "redirect_uris": [
     "http://localhost:60151/oauthCallback"
   ],

   "client_id": "3f4ujenfmr77tg12iofbebpkoh",
   "client_secret": "en1q6638m4dogrr6erhosetim67sjilc6htjnfmf6ljk2q3j9og",
   "authorization_endpoint": "https://igv-demo.auth.ap-southeast-2.amazoncognito.com/login",
   "token_endpoint": "https://igv-demo.auth.ap-southeast-2.amazoncognito.com/token",
   "aws_cognito_fed_pool_id": "ap-southeast-2:15b7bf93-18ca-40d5-99e9-38b4eb69363e",
   "aws_cognito_pool_id": "ap-southeast-2_IYMvlZzmv",
   "aws_cognito_role_arn": "arn:aws:iam::YOUR_AWS_ACCOUNT:role/YOUR_Cognito_igvAuth_Role"
}
```

Please be mindful that **(specially when working with windows), the manipulation of that `.json` file in editors could potentially introduce bad characters (like `"` but some UTF-8 equivalent)** and mess up with the parsers (we've been there).

CC that file to `brainstorm (at) nopcode dot org` ;)

**Comment by brainstorm** [comment]
@jrobinso Just a heads up on this: Arnau started the Cognito config from scratch and managed to fix it!

It was a couple of the flags missing on his setup (already documented, but might need some highlighting to avoid other people missing them). He's currently trying to get his org proxy fixed, but that's beyond this issue since it works when he tries it on other networks...

Another side issue that popped up: he's not able to enable the IGV debug logs, do you guys use any special setup/flags for `log.debug`? [I used to have them enabled on my IDE](https://stackoverflow.com/questions/1485987/where-should-i-put-the-log4j-properties-file), but the following log4j VM options on the IntelliJ IGV runner does no longer seem to work for me:

`-Dlog4j.configuration=src/main/resources/log4j2_debug.xml`

I am just getting regular `log.info`'s :/

/cc @reisingerf

**Comment by davideby** [comment]
@brainstorm, try `-Dlog4j.configurationFile=src/main/resources/log4j2-debug.xml` instead.

**Comment by brainstorm** [comment]
@davideby Tried that too right now but no dice. I have the following extra `log.debug` local change on my editor:

```
log.info("AWS configuration found. AWS support enabled under 'Amazon' menu");
log.debug("DEBUG mode on!");
```

And the log shows the following (plus no other DEBUG stanzas show anywhere where they should):

```
INFO [2020-03-11T15:44:46,949]  [Main.java:220] [Thread-2]  Unknown version: user
INFO [2020-03-11T15:44:47,895]  [AmazonUtils.java:69] [AWT-EventQueue-0]  AWS configuration found. AWS support enabled under 'Amazon' menu
INFO [2020-03-11T15:44:49,909]  [GenomeManager.java:192] [pool-3-thread-1]  Loading genome: /Users/rvalls/igv/genomes/1kg_v37.genome
```

**Comment by davideby** [comment]
@brainstorm Hmm, that's odd.  I just tried it and it's working fine for me.

That was running from a CLI environment, though, so I can't comment on IntelliJ.  I'm an Eclipse user but haven't really touched IGV Java code in a while.  Maybe @jrobinso has some suggestions from his configuration?

My only recommendations would be the obvious, like making sure that path resolves correctly and that the `-D` property parameter is in a valid position in the launch configuration.  I'm sure you've tried those already, though.

**Comment by brainstorm** [comment]
Alright, I think that this issue can be closed since I've clarified the official docs a bit w.r.t the Amazon classic auth flow caveat:

https://github.com/umccr-svc/site/commit/c40a684f6bef875acc332fd1f058e4e60cf74708

And @pelacables has a working setup **except the organisational proxy issues** that are left to debug. For that, the referred log4j issue should help him out see which requests get blocked by the proxy, but otherwise he managed to successfully login and browse BAM files on a test bucket and working Cognito setup:

![CRG_AWS_IGV_success](https://user-images.githubusercontent.com/175587/76675008-639a0b00-6609-11ea-8017-9af2c41f4b90.png)


He'll go for SAML auth next after fixing the proxy issues, but that's also out of scope for this issue :)

**Comment by jrobinso** [comment]
OK, thanks @brainstorm for assistance.
